lib/galaxy/tools/__init__.py

```diff
@@ -3487,6 +3487,16 @@ def element_is_valid(element: model.DatasetCollectionElement):
         return element.element_object.is_ok
 
 
+class KeepSuccessDatasetsTool(FilterDatasetsTool):
+    tool_type = "keep_success_datasets_collection"
+    require_terminal_states = False
+    require_dataset_ok = False
+
+    @staticmethod
+    def element_is_valid(element: model.DatasetCollectionElement):
+        return element.element_object.is_ok
```

Comment:
I think that will also skip new/queued/running etc ...

Reply:
I think it is ok, no?
Someone could be interested in getting the valid datasets while other are still running.

Reply:
I don't like that this is not deterministic. If you extract a workflow form this it will behave very differently as when you ran this tool manually. You can always drag the successful element out if needed ?

Reply:
Yes, but you loose the label....
But I agree with your point. I will modify it.

Reply:
> Yes, but you loose the label....

that's a bug we should've fixed ages ago ... I'm sure we'll get to it eventually.

Reply:
Ideally I could create another tool that would be 'extract multiple datasets from collection' that does not require a terminal state and select multiple datasets per index... This way we keep the label.

Reply:
This tool would also allow to filter datasets which are too long to run, in paused state etc...

Reply:
This should be done with an external script or maybe some new option in the UI (pull out finished datasets). A tool is not how we would normally we want to do this - the collection operation tools  were implemented as tools so they can be extracted into workflows. As @mvdbeek - it wouldn't make sense to extract this operation so it should just be implemented with the API IMO.